sqlite/src/dttz.c

```diff
@@ -153,134 +157,130 @@ int convMem2ClientDatetimeStr(Mem *pMem, void *out, int outlen, int *outdtsz) {
-                sqlite3_result_null(context);
-                return;
-            }
-
-            sqlite3_result_int(context, cdt.tm.tm_mon+1);
-      }
-
-  } else 
-      sqlite3_result_null(context);
+  datetimeComponentFunc(context, argc, argv, tm.tm_mon+1);
 }
 
 static void dayFunc(sqlite3_context *context, int argc, sqlite3_value **argv) {
-
-  assert(argc == 1);
-
-  if( SQLITE_NULL!=sqlite3_value_type(argv[0]) ){
-
-      if(sqlite3VdbeMemDatetimefy(argv[0]) == SQLITE_OK) {
-            cdb2_client_datetime_t   cdt;
-
-            if(convMem2ClientDatetime(argv[0], &cdt) != SQLITE_OK) {
-                sqlite3_result_null(context);
-                return;
-            }
-
-            sqlite3_result_int(context, cdt.tm.tm_mday);
-      }
-
-  } else 
-      sqlite3_result_null(context);
+  datetimeComponentFunc(context, argc, argv, tm.tm_mday);
 }
 
 static void hourFunc(sqlite3_context *context, int argc, sqlite3_value **argv) {
-
-  assert(argc == 1);
-
-  if( SQLITE_NULL!=sqlite3_value_type(argv[0]) ){
-
-      if(sqlite3VdbeMemDatetimefy(argv[0]) == SQLITE_OK) {
-            cdb2_client_datetime_t   cdt;
-
-            if(convMem2ClientDatetime(argv[0], &cdt) != SQLITE_OK) {
-                sqlite3_result_null(context);
-                return;
-            }
-
-            sqlite3_result_int(context, cdt.tm.tm_hour);
-      }
-
-  } else 
-      sqlite3_result_null(context);
+  datetimeComponentFunc(context, argc, argv, tm.tm_hour);
 }
 
 static void minuteFunc(sqlite3_context *context, int argc, sqlite3_value **argv) {
+  datetimeComponentFunc(context, argc, argv, tm.tm_min);
+}
+
+static void secondFunc(sqlite3_context *context, int argc, sqlite3_value **argv) {
+  datetimeComponentFunc(context, argc, argv, tm.tm_sec);
+}
 
+static void milliSecondFunc(sqlite3_context *context, int argc, sqlite3_value **argv) {
   assert(argc == 1);
 
   if( SQLITE_NULL!=sqlite3_value_type(argv[0]) ){
+      Mem *pMem = argv[0];
 
       if(sqlite3VdbeMemDatetimefy(argv[0]) == SQLITE_OK) {
-            cdb2_client_datetime_t   cdt;
-
-            if(convMem2ClientDatetime(argv[0], &cdt) != SQLITE_OK) {
-                sqlite3_result_null(context);
-                return;
-            }
-
-            sqlite3_result_int(context, cdt.tm.tm_min);
+          cdb2_client_datetime_t cdt;
+          cdb2_client_datetimeus_t cdtus;
+          void *out = NULL;
+          switch (pMem->du.dt.dttz_prec) {
+              case DTTZ_PREC_MSEC:
+                  out = &cdt;
+                  break;
+              case DTTZ_PREC_USEC:
+                  out = &cdtus;
+                  break;
+          }
+          assert(out != NULL);
+
+          if(convMem2ClientDatetime(argv[0], out) != SQLITE_OK) {
+              sqlite3_result_null(context);
+              return;
+          }
+
+          switch (pMem->du.dt.dttz_prec) {
+              case DTTZ_PREC_MSEC:
+                  sqlite3_result_double(context, (double) cdt.msec);
```

Comment:
I understand why you would want to return a double here. However it seems a bit inconsistent if the fraction field is returned as a double but the seconds field is returned as an integer. After all both "1.2 second" and "0.1 milliseconds" are valid. Maybe just round it down here. What do you think?

Reply:
I thought about that, and yes, it's inconsistent with the other fields.   The problem is that we truncate data - microseconds(....7us) -> 0, which also felt wrong.  It makes sense that the lowest resolution field supports decimals.

Reply:
Another option is to return the milliseconds/microseconds as if they were two separate fields.  Ie: 2020-01-01T01:01:01.002004 returns 2 for millisecond() and 4 for microseconds.  That's consistent with other fields.

Reply:
I rethought about it - you're right it makes more sense to return decimals for fractions.

Reply:
It *is* a bit odd that for '2021-02-03T04:05:06.000007', millisecond() returns 0.007000. Not sure if any alternative is better.
For, microsecond(), there will never be digits beyond the decimal point, right? So microsecond(b)=7.000000 could in fact be an integer? (I'm not sure i understand Mike's first comment about 7us -> 0.)